src/punycode.cpp

```diff
@@ -0,0 +1,338 @@
+            if (value < 0x80)
+            {
+                // copy them to the output in order
+                output.append(1, static_cast<char>(value));
+            }
+            codepoints.push_back(value);
+        }
+
+        // let h = b = the number of basic code points in the input
+        size_t h = output.size();
+        size_t b = h;
+
+        // copy a delimiter if b > 0
+        if (b > 0)
+        {
+            output.append(1, '-');
+        }
+    
+        // while h < length(input) do begin
+        while (h < codepoints.size())
+        {
+            // let m = the minimum {non-basic} code point >= n in the input
+            long m = std::numeric_limits<long>::max();
+            for (auto it = codepoints.begin(); it != codepoints.end(); ++it)
+            {
+                if ((*it >= n) && (*it < m))
+                {
+                    m = *it;
+                }
+            }
+
+            // let delta = delta + (m - n) * (h + 1), fail on overflow
+            delta += (m - n) * (h + 1);
+            if (delta < 0)
+            {
+                throw std::invalid_argument("Overflow delta");
+            }
+
+            // let n = m    
+            n = m;
+
+            // for each code point c in the input (in order) do begin
+            for (auto it = codepoints.begin(); it != codepoints.end(); ++it)
+            {
+                // if c < n {or c is basic} then increment delta, fail on overflow
+                if (*it < n)
+                {
+                    ++delta;
+                }
+                if (delta < 0)
+                {
+                    throw std::invalid_argument("Overflow delta 2");
+                }
+
+                // if c == n then begin
+                if (*it == n)
+                {
+                    // let q = delta
+                    long q = delta;
+
+                    // for k = base to infinity in steps of base do begin
+                    for (long k = BASE; ; k += BASE)
+                    {
+                        // let t = tmin if k <= bias {+ tmin}, or
+                        //         tmax if k >= bias + tmax, or k - bias otherwise
+                        long t = k - bias;
+                        if (k <= bias)
+                        {
+                            t = TMIN;
+                        }
+                        else if (k >= (bias + TMAX))
+                        {
+                            t = TMAX;
+                        }
+                        
+                        // if q < t then break
+                        if (q < t)
+                        {
+                            break;
+                        }
+
+                        // output the code point for digit t + ((q - t) mod (base - t))
+                        output.append(1, DIGIT_TO_BASIC[t + ((q - t) % (BASE - t))]);
+
+                        // let q = (q - t) div (base - t)
+                        q = (q - t) / (BASE - t);
```

Comment:
Unless the perf improvement is substantial, I'm trying to follow the pseudocode and example code closely so that it's easy to follow when looking back at the RFC.